### PR TITLE
PCHR-3103: Multi Credential Support for Backstop JS

### PR DESCRIPTION
**Pull request by deb1990**
## Overview
This PR adds the ability to specify different login credentials for different Backstop JS scenarios.

## Problem
Previously we could specify a single credential for the entire Backstop JS suite. But different scenarios needed different login credentials. For example: In L&A we have scenarios for `staff`, `manager` or `admin`. So it was never possible to execute all scenarios at one go.

## Solution
1. In `gulpfile.js`, changed the credentials template to 
```
'credentials': {
    'admin': { 'name': '%{admin-name}', 'pass': '%{admin-password}' },
    'manager': { 'name': '%{manager-name}', 'pass': '%{manager-password}' },
    'staff': { 'name': '%{staff-name}', 'pass': '%{staff-password}' }
 }
```

So now its possible to specify login credentials for all type of roles.

2. If the credentials for one scenario is different than the previous one, then it is necessary to logout and re login. To implement that

In `gulpfile.js`, the login script is being called again different credentials are detected.
In `login.js`, script has been added to logout before logging in again.

3. If a scenario needs a different logon information than the default one, it needs to be specified in the scenario JSON itself, example
```
{
  "scenarios": [
    {
      "label": "<label>",
      "url": "<url>",
      "onReadyScript": "<file name>",
      "credential": "<manager/staff/etc>"
    }
  ]
}
```
`ssp-la-manager-leave-calendar.json` and `ssp-la-my-report.json` is updated to show how we need to specify the credential information.

4. Backstop will use `admin` logon information as default, if no `credential` is mentioned in the scenario.

5. If you have a .DS_Store file or anything else other than a `.json` file in the folder, the code tries to read that file too and the tasks stop without any error. To prevent that the following code has been added `scenario.endsWith('.json')` at

https://github.com/civicrm/civihr/pull/2378/files/88ef00c200ff8b9d73ee13624e305d1fb2998707#diff-cd91b40c4aa1a3aa35c573b56c5cce09R140

## Comments
Part of this feature was developed in https://github.com/civicrm/civihr/pull/1785, which was still in progress.

### Removed Scenario
`SSP - Manager Absence Approval`, `SSP - Edit Emergency Contact` and related files has been removed because both these pages does not exist anymore.

### Failing Scenarios
Some of the missing required data has been updated in https://compucorp.atlassian.net/wiki/spaces/PCHR/pages/117984407/BackstopJS+CiviHR

All the scenarios has been run, and these following does not run
1. leave-balances.json
2. ssp-la-manager-leave-requests.json
3. ta-documents.json

They need to be fixed in future PRs.
Also note that though some of the scenarios working, it has been noticed that few screenshots are not for what they meant to be, so a through check is required in future.

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/login.js [hunk 1]
```diff
@@ -1,11 +1,23 @@
...
   var loginFormSelector = 'form#user-login-form';
+  var credentials = config.credentials[scenario.credential || 'admin'];
 
   casper
-    .echo('Logging in before starting...', 'INFO')
+    .then(function () {
+      if (casper.exists('a[href="/user/logout"]')) {
+        casper.echo('Current scenario has different login credentials from previous, logging out is necessary', 'INFO');
+        casper.echo('Logging Out', 'INFO');
+        return casper.evaluate(function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/login.js]: what if you simply do `phantom.clearCookies();` see here http://darrendev.blogspot.sk/2014/03/clearing-cookies-between-casperjs-tests.html

Comment (deb1990) [reply uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/login.js]: Yes it works, thanks for the suggestion. So I will take the ticket out of POC and estimate.

File: uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/login.js [hunk 2]
```diff
@@ -1,11 +1,22 @@
...
-module.exports = function (casper) {
+module.exports = function (casper, scenario) {
   var config = require('../site-config');
   var loginFormSelector = 'form#user-login-form';
+  var credentials = config.credentials[scenario.credential || 'admin'];
 
   casper
-    .echo('Logging in before starting...', 'INFO')
+    .then(function () {
+      if (casper.exists('a[href="/user/logout"]')) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/login.js]: you are not using the link anymore, so this check is not necessary

File: uk.co.compucorp.civicrm.hrcore/gulpfile.js [hunk 3]
```diff
@@ -128,12 +132,12 @@ var Promise = require('es6-promise').Promise;
...
    * @return {Array}
    */
-  function scenariosList() {
+  function scenariosList () {
     var scenariosPath = backstopDir + 'scenarios/';
 
     return _(fs.readdirSync(scenariosPath))
       .filter(function (scenario) {
-        return argv.configFile ? scenario === argv.configFile : true;
+        return argv.configFile ? scenario === argv.configFile : true && scenario.endsWith('.json');
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/gulpfile.js]: why is this additional check necessary?

File: uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/login.js [hunk 4]
```diff
@@ -1,11 +1,22 @@
 'use strict';
 
-module.exports = function (casper) {
+module.exports = function (casper, scenario) {
   var config = require('../site-config');
   var loginFormSelector = 'form#user-login-form';
+  var credentials = config.credentials[scenario.credential || 'admin'];
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/login.js]: i don't think this default logic should be here, the default value should be defined in the gulpfile. This script should only care that there is a credential definition for the scenario, and nothing else

File: uk.co.compucorp.civicrm.hrcore/gulpfile.js [hunk 5]
```diff
@@ -143,7 +147,15 @@ var Promise = require('es6-promise').Promise;
         return _.assign(scenario, { delay: scenario.delay || 6000 });
       })
       .tap(function (scenarios) {
-        scenarios[0].onBeforeScript = "login";
+        var previousCredential = null;
+
+        scenarios.forEach(function (scenario) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/gulpfile.js]:
```js
scenarios.forEach(function (scenario, index) {
  if (index === 0 || previousCredential !== scenario.credential) {
    scenario.onBeforeScript = 'login';
  }
});

Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/gulpfile.js]: also, i think there should be some logic here that sets the default (admin) credential here if there is none defined

File: uk.co.compucorp.civicrm.hrcore/gulpfile.js (hunk 3 above)
Comment (deb1990) [reply uk.co.compucorp.civicrm.hrcore/gulpfile.js]: If you have a .DS_Store file or anything else other than a `.json` file in the folder, the code tries to read that file too and the tasks stop without any error. This is prevent that. 
Sorry, Forgot to mention this in the doc, updating.

File: uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/login.js (hunk 2 above)
Comment (deb1990) [reply uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/login.js]: true, but still we need to check whether logout is necessary, so edited this part of the code.

**Comment by deb1990** [comment]
> Figure out what the problem is, at least at a very superficial level. Saying "there is a problem, should be figure out" doesn't help anyone.

@AkA84 As I did not have more time approval from Mike to spend on this ticket, the plan was to find out the faulty scenarios as part of this PR, and then fix them in the next PRs. I was also going to send an email about this amongst fellow front end devs. 

Does that sound good?

File: uk.co.compucorp.civicrm.hrcore/gulpfile.js [hunk 6]
```diff
@@ -147,11 +147,15 @@ var Promise = require('es6-promise').Promise;
         return _.assign(scenario, { delay: scenario.delay || 6000 });
       })
       .tap(function (scenarios) {
-        var previousCredential = null;
+        var DEFAULT_CREDENTIAL = 'admin';
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/gulpfile.js]: put this at the top of the IIFE, among the other variables

File: uk.co.compucorp.civicrm.hrcore/gulpfile.js [hunk 7]
```diff
@@ -147,11 +147,15 @@ var Promise = require('es6-promise').Promise;
...
+        var previousCredential;
 
-        scenarios.forEach(function (scenario) {
-          if (previousCredential !== scenario.credential || previousCredential === null) {
+        scenarios.forEach(function (scenario, index) {
+          scenario.credential = scenario.credential || DEFAULT_CREDENTIAL;
+
+          if (index === 0 || previousCredential !== scenario.credential) {
             scenario.onBeforeScript = 'login';
+            scenario.isLogOutNecessary = true;
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/gulpfile.js]: `scenario.performLogout`

Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/gulpfile.js]: also, you need to set it to `false` if it's not `true`

Comment (deb1990) [reply uk.co.compucorp.civicrm.hrcore/gulpfile.js]: As discussed, not adding it.

File: uk.co.compucorp.civicrm.hrcore/gulpfile.js [hunk 8]
```diff
@@ -7,6 +7,7 @@ var file = require('gulp-file');
 var backstopjs = require('backstopjs');
 var fs = require('fs');
 var Promise = require('es6-promise').Promise;
+var DEFAULT_CREDENTIAL = 'admin';
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/gulpfile.js]: i said the IIFEE

```js
// BackstopJS tasks
(function () {

}());

Comment (deb1990) [reply uk.co.compucorp.civicrm.hrcore/gulpfile.js]: Oops, fixed.

**Pull request merged**